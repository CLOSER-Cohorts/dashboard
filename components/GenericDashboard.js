import DataTable from './DataTable';
import Tabs from '@mui/material/Tabs';
import Tab from '@mui/material/Tab';
import TabPanel from './TabPanel';
import Box from '@mui/material/Box';

export default function GenericDashboard(props) {

  function a11yProps(index) {
    return {
      id: `simple-tab-${index}`,
      'aria-controls': `simple-tabpanel-${index}`,
    };
  }

  function dashboardTabs() {
    return props.tabNames.map((tabName, i) => {
      return <Tab label={tabName} style={{ fontSize: '.6em', width: 'min-content', textAlign: 'justify' }} {...a11yProps(i)} key={i} />
    })

  }

  function dashboardTabPanels() {
    return Object.keys(props.tableData).map((tabName, i) => {
      return <TabPanel value={props.value} index={i} key={i}>
        <div key={0}>
          <DataTable key={0} data={props.tableData}
            allData={props.data}
            headers={[tabName]}
            updateDetailsPanel={props.updateSelectedValueDetails}
            colecticaRepositoryHostname={props.colecticaRepositoryHostname}
            panelContents={props.panelContents}
          />
        </div>
      </TabPanel>
    })

  }

  return <Box sx={{ width: '100%' }}>
    <Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
      <Tabs value={props.value} 
        onChange={props.handleChange} 
        aria-label="Dashboard tabs"
        sx={{
          '.MuiTabs-flexContainer': {
            flexWrap: 'wrap',
          },
        }}
        >
        {dashboardTabs()}
      </Tabs>
    </Box>
    <div style={{ "display": "flex", "flexDirection": "row" }}>
      <div style={{ "width": "60%" }}>
        {dashboardTabPanels()}
      </div>
      <div style={{ "flex": "1 1 100%" }}>{props.selectedValueDetails}</div>
    </div>
  </Box>

}