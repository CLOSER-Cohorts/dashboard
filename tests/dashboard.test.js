import renderer from 'react-test-renderer';

import GenericDashboard from '../components/GenericDashboard';

const colecticaQueryResults = {
  'Lifestage': [{
    agency: 'uk.lha',
    userAttributeValue: 'Third age',
    studyUnitIdentifier: '11fc8ab8-e8ff-4364-bf18-0a99e36b048b'
  },
  {
    agency: 'uk.cls.nextsteps',
    userAttributeValue: 'Adolescence',
    studyUnitIdentifier: '487eae00-70da-46fc-a2e4-c8eb081c5b8f'
  },
  {
    agency: 'uk.cls.nextsteps',
    userAttributeValue: 'Adolescence',
    studyUnitIdentifier: '92f50262-1a83-4ad0-bfb4-6129a0a8d7ad'
  }]
}

const panelContents = (tableCell, e, data, tableHeaders, hostname) => {

  const selectedFieldValueInstances = data[tableHeaders[0]].filter(
    ddiElement => (Object.keys(ddiElement).includes('userAttributeValue')
      ? ddiElement.userAttributeValue : ddiElement) === e.currentTarget.textContent.replace("\u00A0", ' '))

  return <div><h2>{tableCell}</h2>
    <ul>
      {selectedFieldValueInstances.map(selectedFieldInstance => {
        const url = `https://${hostname}/item/${selectedFieldInstance.agency}/${selectedFieldInstance.studyUnitIdentifier}`
        return <li><a target="_blank" href={url}>{url}</a></li>
      }
      )}
    </ul>
  </div>

}

function getTableData(colecticaQueryResults) {
  var tableData = {}
  !colecticaQueryResults.ErrorMessage ? Object.keys(colecticaQueryResults).map((dataField, index) => {
    const uniqueValues = colecticaQueryResults[dataField] ? [...new Set(colecticaQueryResults[dataField].map(data => {
      return Object.keys(data).includes('userAttributeValue') ? data.userAttributeValue : data;

    }))].sort((x, y) => x.toString().charCodeAt() - y.toString().charCodeAt()) : []

    tableData[dataField] = uniqueValues.map(uniqueValue => {
      return !!uniqueValue && [uniqueValue.replace(' ', "\u00A0"), colecticaQueryResults[dataField]
        .filter(
          fieldValue => (Object.keys(fieldValue).includes('userAttributeValue')
            ? fieldValue.userAttributeValue
            : fieldValue) === uniqueValue).length]
    })
  }) : `Error retrieving data, the Collectica API is not reachable. ${colecticaQueryResults.ErrorMessage}`
  return tableData
}

const tabNames = ["Lifestage",
  "Lifestage Description",
  "Creator",
  "Publisher",
  "Analysis Unit",
  "Kind Of Data",
  "Country",
  "Mode of Collection",
  "Type of Mode of Collection"
];

const tableData = getTableData(colecticaQueryResults)

it('renders correctly', () => {
  const tree = renderer
    .create(<GenericDashboard value={0}
      data={colecticaQueryResults}
      tabNames={tabNames}
      panelContents={panelContents}
      tableData={tableData}
    />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});