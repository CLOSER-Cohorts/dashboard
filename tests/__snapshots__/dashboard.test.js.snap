// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="MuiBox-root css-8atqhb"
>
  <div
    className="MuiBox-root css-1gsv261"
  >
    <div
      className="MuiTabs-root css-yiej45-MuiTabs-root"
    >
      <div
        className="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
        style={
          {
            "marginBottom": -0,
            "overflow": "hidden",
          }
        }
      >
        <div
          aria-label="Dashboard tabs"
          aria-orientation={null}
          className="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
          onKeyDown={[Function]}
          role="tablist"
        >
          <button
            aria-controls="simple-tabpanel-0"
            aria-selected={true}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            disabled={false}
            id="simple-tab-0"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            style={
              {
                "fontSize": ".6em",
                "textAlign": "justify",
                "width": "min-content",
              }
            }
            tabIndex={0}
            type="button"
          >
            Lifestage
            <span
              className="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
              style={{}}
            />
          </button>
          <button
            aria-controls="simple-tabpanel-1"
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            disabled={false}
            id="simple-tab-1"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            style={
              {
                "fontSize": ".6em",
                "textAlign": "justify",
                "width": "min-content",
              }
            }
            tabIndex={-1}
            type="button"
          >
            Lifestage Description
          </button>
          <button
            aria-controls="simple-tabpanel-2"
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            disabled={false}
            id="simple-tab-2"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            style={
              {
                "fontSize": ".6em",
                "textAlign": "justify",
                "width": "min-content",
              }
            }
            tabIndex={-1}
            type="button"
          >
            Creator
          </button>
          <button
            aria-controls="simple-tabpanel-3"
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            disabled={false}
            id="simple-tab-3"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            style={
              {
                "fontSize": ".6em",
                "textAlign": "justify",
                "width": "min-content",
              }
            }
            tabIndex={-1}
            type="button"
          >
            Publisher
          </button>
          <button
            aria-controls="simple-tabpanel-4"
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            disabled={false}
            id="simple-tab-4"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            style={
              {
                "fontSize": ".6em",
                "textAlign": "justify",
                "width": "min-content",
              }
            }
            tabIndex={-1}
            type="button"
          >
            Analysis Unit
          </button>
          <button
            aria-controls="simple-tabpanel-5"
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            disabled={false}
            id="simple-tab-5"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            style={
              {
                "fontSize": ".6em",
                "textAlign": "justify",
                "width": "min-content",
              }
            }
            tabIndex={-1}
            type="button"
          >
            Kind Of Data
          </button>
          <button
            aria-controls="simple-tabpanel-6"
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            disabled={false}
            id="simple-tab-6"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            style={
              {
                "fontSize": ".6em",
                "textAlign": "justify",
                "width": "min-content",
              }
            }
            tabIndex={-1}
            type="button"
          >
            Country
          </button>
          <button
            aria-controls="simple-tabpanel-7"
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            disabled={false}
            id="simple-tab-7"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            style={
              {
                "fontSize": ".6em",
                "textAlign": "justify",
                "width": "min-content",
              }
            }
            tabIndex={-1}
            type="button"
          >
            Mode of Collection
          </button>
          <button
            aria-controls="simple-tabpanel-8"
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            disabled={false}
            id="simple-tab-8"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            style={
              {
                "fontSize": ".6em",
                "textAlign": "justify",
                "width": "min-content",
              }
            }
            tabIndex={-1}
            type="button"
          >
            Type of Mode of Collection
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    style={
      {
        "display": "flex",
        "flexDirection": "row",
      }
    }
  >
    <div
      style={
        {
          "width": "60%",
        }
      }
    >
      <div
        aria-labelledby="simple-tab-0"
        hidden={false}
        id="simple-tabpanel-0"
        role="tabpanel"
      >
        <div
          className="MuiBox-root css-19kzrtu"
        >
          <div>
            <table
              table-layout="fixed"
              width="100%"
            >
              <thead>
                <tr>
                  <th
                    align="left"
                  >
                    Lifestage
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td
                    align="left"
                    onClick={[Function]}
                    width="80%"
                  >
                    <a
                      role="button"
                      tabIndex="0"
                      title="Click here to view links to items containing 'Adolescence'"
                    >
                      Adolescence
                    </a>
                  </td>
                  <td
                    align="left"
                    onClick={null}
                    width="20%"
                  >
                    2
                  </td>
                </tr>
                <tr>
                  <td
                    align="left"
                    onClick={[Function]}
                    width="80%"
                  >
                    <a
                      role="button"
                      tabIndex="0"
                      title="Click here to view links to items containing 'Third age'"
                    >
                      Third age
                    </a>
                  </td>
                  <td
                    align="left"
                    onClick={null}
                    width="20%"
                  >
                    1
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
    <div
      style={
        {
          "flex": "1 1 100%",
        }
      }
    />
  </div>
</div>
`;
